### PR TITLE
feat(arns): add API to block and unblock ArNS names

### PR DESCRIPTION
**Pull request by karlprieb**

### CODE REVIEW

File: src/middleware/arns.ts [hunk 1]
```diff
@@ -108,3 +117,81 @@ export const createArnsMiddleware = ({
     res.header('Cache-Control', `public, max-age=${ttl}`);
     dataHandler(req, res, next);
   });
+
+export class BlockedNamesCache {
```
Comment (karlprieb) [inline src/middleware/arns.ts:121]: @djwhitt Do you think this is the best place to have this class?

I have some questions about this class:
 1. If the cache is updated from the db when it starts and names are added and removed when we block and unblock names I don't think we need TTL and the setInterval to update the cache from the db.
 2. If 1 is correct we don't even need this class as the cache can be just a variable.

File: src/routes/ar-io.ts [hunk 2]
```diff
@@ -133,6 +133,46 @@ arIoRouter.put('/ar-io/admin/block-data', express.json(), async (req, res) => {
...
+  async (req, res) => {
+    try {
+      const { name } = req.body;
+      if (name === undefined) {
+        res.status(400).send("Must provide 'name'");
+        return;
+      }
+
+      system.db.unblockName({ name });
+      system.blockedNamesCache.removeName(name);
```
Comment (karlprieb) [inline src/routes/ar-io.ts:167]: Should we just run `db.unblockName` from the cache itself and just call `removeName` here?

File: src/routes/ar-io.ts [hunk 3]
```diff
@@ -133,6 +133,46 @@ arIoRouter.put('/ar-io/admin/block-data', express.json(), async (req, res) => {
...
+    }
+
+    system.db.blockName({ name, source, notes });
+    system.blockedNamesCache.addName(name);
+
+    res.json({ message: 'Name blocked' });
+  } catch (error: any) {
+    res.status(500).send(error?.message);
+  }
+});
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/routes/ar-io.ts:152]: Great suggestion!

File: src/routes/ar-io.ts [hunk 4]
```diff
@@ -133,6 +133,46 @@
...
+arIoRouter.put('/ar-io/admin/block-name', express.json(), async (req, res) => {
+  try {
+    const { name, source, notes } = req.body;
+    if (name === undefined) {
+      res.status(400).send("Must provide 'name'");
+      return;
+    }
+
+    system.db.blockName({ name, source, notes });
+    system.blockedNamesCache.addName(name);
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/routes/ar-io.ts:146]: @djwhitt this is a real issue. Maybe we should wait for the db response to update the cache and to respond the api request

[the remark this answers is not in the document]
Comment (djwhitt) [reply src/routes/ar-io.ts:146]: Agreed. We should wait on the DB operation to complete before updating the cache.

File: test/moderation-schema.sql [hunk 5]
```diff
@@ -20,3 +20,9 @@ CREATE TABLE blocked_hashes (
 CREATE INDEX blocked_hashes_source_id_idx
   ON blocked_hashes (block_source_id);
 CREATE UNIQUE INDEX block_sources_name_idx ON block_sources (name);
+CREATE TABLE blocked_names (
+  name TEXT PRIMARY KEY,
+  block_source_id INTEGER,
+  notes TEXT,
+  blocked_at INTEGER
+);
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply test/moderation-schema.sql:28]: @karlprieb We don't need the foreign key, but we should add the index.

File: src/routes/ar-io.ts (hunk 2 above)
Comment (djwhitt) [reply src/routes/ar-io.ts:167]: I lean towards keeping them separate. That keeps the cache interface simpler.

File: src/routes/ar-io.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/routes/ar-io.ts:152]: Agreed, let's do this.

File: src/middleware/arns.ts (hunk 1 above)
Comment (karlprieb) [reply src/middleware/arns.ts]: [`95f516a` (#245)](https://github.com/ar-io/ar-io-node/pull/245/commits/95f516a24ea4498d67addb2668035d6f02c17d96)

File: src/routes/ar-io.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/routes/ar-io.ts]: [`2f02a4a` (#245)](https://github.com/ar-io/ar-io-node/pull/245/commits/2f02a4a48323fccab7e79a6e4804bb599501c594)

File: test/moderation-schema.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (karlprieb) [reply test/moderation-schema.sql:28]: [`bf32fd2` (#245)](https://github.com/ar-io/ar-io-node/pull/245/commits/bf32fd2210ae5af5e164d00097110c632af9b43b)

**Review by djwhitt — APPROVED** [review]

**Pull request merged**